### PR TITLE
refactor: data validation using schema

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
- Add `zod` schema for `monorepo.json` which will be used during linting for validation
- Add JSON schema for `monorepo.json` for better user ux

## Context
Ref: https://github.com/renovatebot/renovate/issues/29942#issuecomment-2241499559

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: tools/schemas/monorepo-schema.json
Comment (viceice) [inline tools/schemas/monorepo-schema.json:11]: do we really need this? 🤔

File: tools/validate-schemas.ts
Comment (viceice) [inline tools/validate-schemas.ts:7]: i would prefer to use jest test for better output on PR's

File: tools/schemas/monorepo-schema.json
Comment (RahulGautamSingh) [reply tools/schemas/monorepo-schema.json:11]: I get a minor wraning from lint when this is not added.

File: tools/validate-schemas.ts
Comment (RahulGautamSingh) [reply tools/validate-schemas.ts:7]: Okay...adding a test instead

File: tools/schemas/monorepo-schema.json
Comment (viceice) [reply tools/schemas/monorepo-schema.json:11]: use a newer json schema

https://json-schema.org/specification-links#published-drafts

Comment (RahulGautamSingh) [reply tools/schemas/monorepo-schema.json:10]: We use the same for renovate.json 's schema and it works fine ... 🤔

File: lib/data/monorepo.json
Comment (viceice) [inline lib/data/monorepo.json:354]: this fix should be in a separate PR

File: test/validate-schemas.spec.ts
Comment (viceice) [inline test/validate-schemas.spec.ts:20]:
```suggestion
      const correspondingDataFileName = schemaFile.replace('-schema', '');
```

File: test/validate-schemas.spec.ts
Comment (viceice) [inline test/validate-schemas.spec.ts:28]:
```suggestion
        dataFileName: correspondingDataFileName,
```

File: test/validate-schemas.spec.ts
Comment (viceice) [inline test/validate-schemas.spec.ts:40]: we don't know which file fails 🤔 we should somehow log it, so you know which file failed

File: test/validate-schemas.spec.ts
Comment (viceice) [inline test/validate-schemas.spec.ts:34]: use our JSON Schema helper 
https://github.com/renovatebot/renovate/blob/e1c7e84deea7febc0c420b6ece89690a39a06af6/lib/util/schema-utils.ts#L200

File: lib/data/monorepo.json
Comment (RahulGautamSingh) [reply lib/data/monorepo.json:354]: https://github.com/renovatebot/renovate/pull/30917

**Comment by HonkingGoose** [comment]
I'll let the maintainers review this PR. I don't know enough to help review. 😉

File: test/validate-schemas.spec.ts
Comment (viceice) [inline test/validate-schemas.spec.ts:33]: use all settled, so we wait for all promises to settle, otherwise it will return early and we've lingering promises

File: test/validate-schemas.spec.ts
Comment (viceice) [inline test/validate-schemas.spec.ts:37]: this can fail too on Json typos

File: test/validate-schemas.spec.ts
Comment (RahulGautamSingh) [reply test/validate-schemas.spec.ts:33]: When using all settled the tests pass regardless of the assertions being false :)

File: test/validate-schemas.spec.ts
Comment (viceice) [inline test/validate-schemas.spec.ts:33]:
```suggestion
    const settledPromises = await Promise.allSettled(
```
why any?

Comment (RahulGautamSingh) [reply test/validate-schemas.spec.ts:33]: TS isn't detecting the correct type for this from the respective `.d.ts` file.

The `.d.ts` file defines the return type for `allSettled` as `PromiseFulfilledType | PromiseRejectedType` but TS only infers `PromiseFulfilledType`, due to which I get TS errors when I use the `reason` field, which is defined in `PromiseRejectedType`

Comment (viceice) [reply test/validate-schemas.spec.ts:33]: remove the reason, status compare should be enough

File: test/validate-schemas.spec.ts
Comment (viceice) [inline test/validate-schemas.spec.ts:51]:
```suggestion
        settledPromise: settledPromises[i],
```

File: test/validate-schemas.spec.ts
Comment (viceice) [inline test/validate-schemas.spec.ts:59]:
```suggestion
```

File: test/validate-schemas.spec.ts
Comment (RahulGautamSingh) [reply test/validate-schemas.spec.ts:33]: The reason holds the invalid json OR invalid schema message so its relevant while debugging failed tests

Comment (RahulGautamSingh) [reply test/validate-schemas.spec.ts]: I did a new commit after only removing the `any` type assertion and not TS errors. Maybe it was a one time bug

File: test/validate-schemas.spec.ts
Comment (viceice) [inline test/validate-schemas.spec.ts:12]: move this to test

**Review by viceice — APPROVED** [review]

**Pull request merged**
